drivers/eeprom/eeprom_emulator.c

```diff
@@ -0,0 +1,543 @@
+ * defined in the dts. The flash partition should allow at least two eeprom
+ * pages.
+ *
+ */
+
+#define DT_DRV_COMPAT zephyr_emu_eeprom
+#include <drivers/eeprom.h>
+#include <drivers/flash.h>
+#include <zephyr.h>
+#define LOG_LEVEL CONFIG_EEPROM_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(eeprom_emulator);
+
+#define DEV_CONFIG(dev) ((dev)->config)
+#define DEV_DATA(dev) ((dev)->data)
+
+
+struct eeprom_emu_config {
+	size_t size;
+	bool readonly;
+	size_t page_size;
+	off_t flash_offset;
+	size_t flash_size;
+	uint8_t flash_cbs;
+	const struct device *flash_dev;
+};
+
+struct eeprom_emu_data {
+	off_t write_offset;
+	off_t page_offset;
+	struct k_mutex lock;
+};
+
+/*
+ * basic flash read, only used with offset aligned to flash write block size
+ */
+static void eeprom_emu_flash_read(const struct device *dev, off_t offset,
+				  uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		if (!flash_read(dev_config->flash_dev, offset, blk, len)) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
+}
+
+/*
+ * basic flash write, only used with offset aligned to flash write block size
+ */
+static void eeprom_emu_flash_write(const struct device *dev, off_t offset,
+				   const uint8_t *blk, size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	int rc;
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
+		rc |= flash_write(dev_config->flash_dev, offset, blk, len);
+		rc |= flash_write_protection_set(dev_config->flash_dev, true);
+		if (!rc) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
+}
+
+/*
+ * basic flash erase, only used with offset aligned to flash page and len a
+ * multiple of the flash page size
+ */
+static void eeprom_emu_flash_erase(const struct device *dev, off_t offset,
+				   size_t len)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	int rc;
+
+	if (device_is_ready(dev_config->flash_dev)) {
+		offset += dev_config->flash_offset;
+		rc = flash_write_protection_set(dev_config->flash_dev, false);
+		rc |= flash_erase(dev_config->flash_dev, offset, len);
+		rc |= flash_write_protection_set(dev_config->flash_dev, true);
+		if (!rc) {
+			return;
+		}
+	}
+	__ASSERT(0, "Flash interface failed");
+}
+
+/*
+ * eeprom_emu_page_invalidate: invalidate a page by writing all zeros at the end
+ */
+static void eeprom_emu_page_invalidate(const struct device *dev, off_t offset)
+{
+	const struct eeprom_emu_config *dev_config = DEV_CONFIG(dev);
+	uint8_t buf[dev_config->flash_cbs];
+
+	LOG_DBG("Invalidating page at [0x%tx]", (ptrdiff_t)offset);
+
+	memset(buf, 0x00, sizeof(buf));
+
+	offset += (dev_config->page_size - sizeof(buf));
+	eeprom_emu_flash_write(dev, offset, buf, sizeof(buf));
```

Comment:
The `eeprom_emu_flash_write` neither reports error nor returns error in case of failure which means that the entire  `eeprom_emu_page_invalidate` may be unsuccessful with not information about that fact.